src/data/real/dedekind_cut.lean

```diff
@@ -0,0 +1,403 @@
+    exact ⟨⟨α, hα, hq⟩, h⟩
+  end }
+
+@[simp]
+lemma mem_cSup {A : set (dedekind_cut Q)} {hA : A.nonempty} {hA' : bdd_above A} :
+  p ∈ cSup A hA hA' ↔ ∃ α ∈ A, α ∈ A ∧ p ∈ α :=
+by simp [cSup]
+
+lemma le_cSup (A : set (dedekind_cut Q)) (hA : A.nonempty) (hA' : bdd_above A) (hα : α ∈ A) :
+  α ≤ cSup A hA hA' :=
+λ p hp, by { simp only [mem_cSup, exists_prop, and_self_left], exact ⟨α, hα, hp⟩ }
+
+lemma is_lub_cSup (A : set (dedekind_cut Q)) (hA : A.nonempty) (hA' : bdd_above A) :
+  is_lub A (cSup A hA hA') :=
+begin
+  refine ⟨λ _, le_cSup _ _ _ _, λ δ hδ s hs, _⟩,
+  simp only [mem_cSup, exists_prop, and_self_left] at hs,
+  obtain ⟨α, hα, hs⟩ := hs,
+  exact hδ hα hs
+end
+
+open_locale pointwise
+
+-- Step 4 (A1)
+instance {Q : Type*} [linear_order Q] [add_group Q]
+  [covariant_class Q Q (+) (<)] [covariant_class Q Q (function.swap (+)) (<)] :
+  has_add (dedekind_cut Q) :=
+⟨λ α β, { carrier := α + β,
+  nonempty' := by { obtain ⟨p, hp⟩ := α.nonempty, obtain ⟨q, hq⟩ := β.nonempty,
+    exact ⟨p + q, p, q, hp, hq, rfl⟩ },
+  ne_top' := begin
+    suffices : ∃ (rs : Q), rs ∉ (α : set Q) + β,
+    { contrapose! this, simp [this] },
+    obtain ⟨r', hr'⟩ := α.exists_not_mem,
+    obtain ⟨s', hs'⟩ := β.exists_not_mem,
+    refine ⟨r' + s', _⟩,
+    rintro ⟨r, s, hr, hs, h⟩,
+    have : r + s < r' + s' :=
+      add_lt_add (lt_of_mem_of_not_mem hr hr') (lt_of_mem_of_not_mem hs hs'),
+    exact absurd h this.ne
+  end,
+  is_lt_lower' := begin
+    rintro p q h ⟨r, s, hr, hs, rfl⟩,
+    refine ⟨q - s, s, is_lt_lower _ _ hr, hs, sub_add_cancel _ _⟩,
+    rwa sub_lt_iff_lt_add
+  end,
+  exists_gt' := begin
+    rintros p ⟨r, s, hr, hs, rfl⟩,
+    obtain ⟨t, ht, hrt⟩ := α.exists_gt hr,
+    exact ⟨t + s, ⟨t, s, ht, hs, rfl⟩, add_lt_add_right hrt _⟩
+  end }⟩
+
+lemma mem_add_iff {Q : Type*} [linear_order Q] [add_group Q] [covariant_class Q Q (+) (<)]
+  [covariant_class Q Q (function.swap (+)) (<)] {α β : dedekind_cut Q} {p : Q} :
+  p ∈ α + β ↔ ∃ (r s : Q), r ∈ α ∧ s ∈ β ∧ r + s = p := iff.rfl
+
+-- Step 4 (A2)
+protected lemma add_comm {Q : Type*} [linear_order Q] [add_comm_group Q]
+  [covariant_class Q Q (+) (<)] (α β : dedekind_cut Q) : α + β = β + α :=
+begin
+  ext,
+  simp only [mem_add_iff, and.comm, and.left_comm, mem_carrier],
+  split;
+  { rintro ⟨r, s, rfl, hr, hs⟩,
+    exact ⟨s, r, add_comm _ _, hs, hr⟩ }
+end
+
+-- Step 4 (A3)
+protected lemma add_assoc {Q : Type*} [linear_order Q] [add_group Q] [covariant_class Q Q (+) (<)]
+  [covariant_class Q Q (function.swap (+)) (<)] (α β γ : dedekind_cut Q) :
+  α + β + γ = α + (β + γ) :=
+begin
+  ext,
+  simp only [mem_carrier, mem_add_iff, and.comm, and.assoc, and.left_comm],
+  split,
+  { rintro ⟨_, r, rfl, hr, s, t, rfl, hs, ht⟩,
+    refine ⟨_, _, _, hs, _, _, rfl, ht, hr⟩,
+    rw add_assoc },
+  { rintro ⟨_, r, rfl, hr, s, t, rfl, hs, ht⟩,
+    refine ⟨_, _, _, ht, _, _, rfl, hr, hs⟩,
+    rw add_assoc }
+end
+
+end dedekind_cut
+
+end
+
+namespace dedekind_cut
```

Comment:
Why close the namespace if you're just going to open it again? If you want to introduce new `variables`, you should just make a new `section` instead.